timemachine/fe/geometry.py

```diff
@@ -3,56 +3,106 @@
-        elif hybridization == HybridizationType.SP:
-            return LocalGeometry.G2_LINEAR
-        else:
-            assert 0, "Unknown 2-nbr geometry!"
-    elif len(nbrs) == 3:
-        if hybridization == HybridizationType.SP3:
-            return LocalGeometry.G3_PYRAMIDAL
-        elif hybridization == HybridizationType.SP2:
-            return LocalGeometry.G3_PLANAR
-        else:
-            assert 0, "Unknown 3-nbr geometry"
-    elif len(nbrs) == 4:
-        if hybridization == HybridizationType.SP3:
-            return LocalGeometry.G4_TETRAHEDRAL
-        else:
-            assert 0, "Unknown 4-nbr geometry"
-    else:
-        assert 0, "Too many neighbors"
-
-
-def classify_geometry(mol: Chem.Mol) -> List[LocalGeometry]:
+    cmat = np.zeros((num_atoms, num_atoms))
+    for i, j in bond_idxs:
+        cmat[i][j] = 1
+        cmat[j][i] = 1
+
+    nblist = []
+    for i in range(num_atoms):
+        nbs = []
+        for j in range(num_atoms):
+            if cmat[i][j]:
+                nbs.append(j)
+        nblist.append(nbs)
+
+    return nblist
+
+
+def label_stereo(
+    num_atoms,
+    bond_idxs,
+    angle_idxs,
+    angle_params,
+    improper_idxs,
+    improper_params,
+):
+    # list of list representation
+    nblist = bond_idxs_to_nblist(num_atoms, bond_idxs)
+    atom_geometries = []
+    for atom_idx, atom_nbs in enumerate(nblist):
+        if len(atom_nbs) == 4:
+            atom_geometries.append(LocalGeometry.G4_TETRAHEDRAL)
+        elif len(atom_nbs) == 3:
+            # check for impropers
+            local_geometry = LocalGeometry.G3_PYRAMIDAL
+            # (tbd): should we also check propers as well?
+            # impropers are centered around the first atom.
+            for (i, _, _, _), (k, p, n) in zip(improper_idxs, improper_params):
+                if i == atom_idx and k > 0.0:
+                    local_geometry = LocalGeometry.G3_PLANAR
+                    break
+
+            atom_geometries.append(local_geometry)
+            # atom_stereo_flags.append(is_stereo)
+        elif len(atom_nbs) == 2:
+            # check angle terms:
+            local_geometry = LocalGeometry.G2_KINK
+            for (i, j, k), (_, angle) in zip(angle_idxs, angle_params):
+                if abs(angle) < 0.05:
+                    assert 0
+                ii, kk = atom_nbs[0], atom_nbs[1]
+                if j == atom_idx:
+                    if (i, k) == (ii, kk) or (i, k) == (kk, ii):
+                        if abs(angle - np.pi) < 0.05:
+                            local_geometry = LocalGeometry.G2_LINEAR
+                            break
+            atom_geometries.append(local_geometry)
+        elif len(atom_nbs) == 1:
+            atom_geometries.append(LocalGeometry.G1_TERMINAL)
+        elif len(atom_nbs) == 0:
+            atom_geometries.append(None)
+
+    return atom_geometries
+
+
+def classify_geometry(mol: Chem.Mol, ff: Forcefield = None, core: List[int] = None) -> List[LocalGeometry]:
```

Comment:
Dumb question - what's the rationale for adding `Optional` to something that already has a default parameter? isn't that a bit redundant?

Reply:
In cases where the default is `None`, the default doesn't type-check

Reply:
>  what's the rationale for adding Optional to something that already has a default parameter? isn't that a bit redundant?

It's a bit confusing because "optional" has several meanings, e.g. an argument that need not be provided, or the type `Optional = Union[T, None]` (the latter is the case here).

Without `Optional` here, the default values don't have the correct type (`None` is not an allowed value of `List[int]`, for example) and mypy won't catch potential bugs in the implementation if e.g. we missed handling the case where `ff = None`.

I'm not sure why mypy with our configuration doesn't catch this, but it's flagged in my editor (using pyright):
![image](https://user-images.githubusercontent.com/319411/169898177-2be2d5d6-3691-4716-af52-44b0a3580435.png)

Reply:
fixed

Reply:
added